### ISSUE TITLE
Declarative Migrations with Just a Schema File (no migrations table and directory)

### ISSUE THREAD

**Issue by SanderGi** [SanderGi/sqlite-auto-migrator#3]
One question; Do I really need that "migration" table and "migrations" directory? Is there a way to turn that off? As my CI simply has a sql-file with the latest schema and the CI always compares the "production" database with the schema and changes it, what exactly would I need the migration table for? I take a backup of the database before, so I don't really need any revert functionality. I am not checking in any versions or rollback features in git. I simply want it to compare the .sql-file with the database-file and create a "migration-file" in RAM and then execute it without leaving any traces.

_Originally posted by @buzzy in https://github.com/SanderGi/sqlite-auto-migrator/issues/1#issuecomment-2372377631_

**Comment by SanderGi** [comment]
_TL;DR: the migration files remove ambiguity. We can add a feature that allows leaving no trace by limiting the schema changes to creates+deletes (no renames). We just have to decide how to expose it in the API_

Excellent question. The migration files in the directory and table are not just for reverting or automatically migrating the database when checking out a new branch/commit in git (in fact, this could be achieved without migration files by just diffing the schema and database as you suggest). **They are mainly there to clarify ambiguity between equivalent schemas that represent different data migration changes**. For instance, deleting a column and creating a new column would result in the exact same schema as instead renaming the old column to the new name. However the latter would move over all the data whereas the former would fill in new NULL/default data and lose the old data. Storing migration files allows sqlite-auto-migrator to prompt the user for which data change they intend with the schema change (e.g., a rename or a delete+create) and then save that decision for applying to the remote production database or another developers local database, etc. **If the decision was not stored, it would be possible for the different instances of the database under the same schema to handle the schema changes differently which could lead to unintended side effects in production that do not appear locally**. After all, production is likely to receive many schema changes at once bundled into one big schema change that thus contains more ambiguity than the incremental changes made locally). The reason there is both a table and directory is to track ambiguity in which migration files have been applied. After all, it is possible to have the same schema come about through different data changes so two migration states could have equivalent schemas but represent different manipulations to the data.

That said, it is also possible to resolve the ambiguities declaratively without storing migration state by limiting the operations that can be performed during a migration. **If, for instance, you are okay only allowing creates/deletes and manually handling renames** (e.g. by ssh-ing into your production database to execute sql commands or by re-enabling migration state), **I could add that feature to the library**. I propose to do this by adding a boolean `trackUnambiguousMigrationState` flag to the migrator class config that defaults to true but you can set to false. Then the library would not create any migration table/directory/files unless you prompt it that your change was a rename. It would instead handle any deletes/creates by diffing the schema and database declaratively in RAM. The downside is that it would have to assume any changes not documented in migration files are deletes/creates that have deliberately not been tracked so it would not be able to warn about changes you have made directly to the database and not captured in the schema (it would instead just undo these changes to match the schema). **Happy to hear what your thoughts are on exposing it this way in the API and if you have better ideas**

**Comment by buzzy** [comment]
That actually sounds exactly what I want! I totally understand (and agree) to the logic behind the migration history for larger teams and projects. But in small teams as mine, I try to never make destructive changes as old backend endpoints still needs to function. There are many different versions of the client running at customers that all uses different versions of the backend endpoints. So in my case, I will only let the migration do non-destructive things like adding fields, indexes etc. I have set all destructive things to MANUAL_MIGRATION so that the CI stops and forces me to do those upgrades manually (which happens rarely). Thus, having a flag that you mention would be perfect for me.

Until now, I have done all database changes manually anyways, so I was thinking of creating this "diff and update" thing myself, but then I found your library :)

**Comment by SanderGi** [comment]
I have pushed a version of declarative diffing to main (still have to write tests+documentation and get your input before publishing to npm).

You can try it out with `bun install github:SanderGi/sqlite-auto-migrator` and changing your code to something like the following:
```
import { Migrator } from "sqlite-auto-migrator";
import { config } from "../config/config";

// @ts-ignore
import sessions from "../databases/sessions.sql";
// @ts-ignore
import main from "../databases/main.sql";

export async function updateDatabase() {
  // Exit early if --update-database flag is not present
  if (Bun.argv.indexOf("--update-database") === -1) return;

  try {
    // Session database
    const migrator = new Migrator({
       dbPath: config.auth.database,
       schemaPath: sessions,
       createDBIfMissing: true,
       onlyTrackAmbiguousState: true,
    });
    await migrator.migrate('latest', {
      onRename: Migrator.REQUIRE_MANUAL_MIGRATION,
      onDestructiveChange: Migrator.REQUIRE_MANUAL_MIGRATION,
      onChangedView: Migrator.REQUIRE_MANUAL_MIGRATION,
      onChangedIndex: Migrator.PROCEED,
      onChangedTrigger: Migrator.PROCEED,
    });

    // Main database
    const migrator2 = new Migrator({
       dbPath: config.database,
       schemaPath: main,
       createDBIfMissing: true,
       onlyTrackAmbiguousState: true,
    });
    await migrator2.migrate('latest', {
      onRename: Migrator.REQUIRE_MANUAL_MIGRATION,
      onDestructiveChange: Migrator.REQUIRE_MANUAL_MIGRATION,
      onChangedView: Migrator.REQUIRE_MANUAL_MIGRATION,
      onChangedIndex: Migrator.PROCEED,
      onChangedTrigger: Migrator.PROCEED,
    });
  } catch (e) {
    console.error(e);
    process.exit(1);
  }

  console.log("Databases updated successfully!");

  process.exit(0);
}
```

I renamed the flag to `onlyTrackAmbiguousState` to [avoid negation in the variable name and have the default be false](https://www.serendipidata.com/posts/naming-guidelines-for-boolean-variables). I also added the option to specify how `migrate` should handle the untracked changes (onDestructiveChange, onChangedView, etc.). For your use case, you will no longer need the `make` call since you will manually handle renames. You can add it at any point in the future if you need to make a rename and want to start tracking the associated ambiguous state. Let me know what your thoughts are and if you run into any bugs with this early version of declarative diffing

**Comment by buzzy** [comment]
I'm afraid I still get deadlocks. But the onlyTrackAmbiguousState flag seems to be working perfectly :)

![image](https://github.com/user-attachments/assets/136b5b3c-9b86-470a-b0c0-a8d504b6d503)

The weird part is that it's random. If I delete the databases and run again, it sometimes works.

**Comment by SanderGi** [comment]
Happy to hear the flag is working for your use case. I'll ping you when I've published this functionality to npm. For me, it seems the deadlocks go away when replacing node-sqlite3 with bun:sqlite (#2)

**Comment by SanderGi** [comment]
Version 1.2.0 has been published to NPM and includes the declarative migrations
